lib/iris/tests/test_io_init.py

```diff
@@ -21,29 +21,31 @@
 class TestDecodeUri(tests.IrisTest):
     def test_decode_uri(self):
         tests = {
-            "/data/local/someDir/PP/COLPEX/COLPEX_16a_pj001.pp": (
+            (uri := "/data/local/someDir/PP/COLPEX/COLPEX_16a_pj001.pp"): (
                 "file",
-                "/data/local/someDir/PP/COLPEX/COLPEX_16a_pj001.pp",
+                uri,
             ),
-            r"C:\data\local\someDir\PP\COLPEX\COLPEX_16a_pj001.pp": (
+            (uri := r"C:\data\local\someDir\PP\COLPEX\COLPEX_16a_pj001.pp"): (
                 "file",
-                r"C:\data\local\someDir\PP\COLPEX\COLPEX_16a_pj001.pp",
+                uri,
             ),
-            "file:///data/local/someDir/PP/COLPEX/COLPEX_16a_pj001.pp": (
-                "file",
-                "///data/local/someDir/PP/COLPEX/COLPEX_16a_pj001.pp",
+            (
+                uri := "file:///data/local/someDir/PP/COLPEX/COLPEX_16a_pj001.pp"
+            ): (
+                uri[:4],
+                uri[7:],
```

Comment:
This is changing what the test checks for, rather than just updating what it uses. I guess it's fine given that the test passes though.

Reply:
Interestingly, what I've specified for the expected result matches exactly what is actually returned e.g.,
Given:
```python
        for uri, expected in tests.items():
            print(f"{uri=}")
            print(f"{expected=}")
            actual = iris.io.decode_uri(Path(uri))
            print(f"{actual=}\n")
            self.assertEqual(expected, actual)
```
We get this:
```python
uri='/data/local/someDir/PP/COLPEX/COLPEX_16a_pj001.pp'
expected=('file', '/data/local/someDir/PP/COLPEX/COLPEX_16a_pj001.pp')
actual=('file', '/data/local/someDir/PP/COLPEX/COLPEX_16a_pj001.pp')

uri='C:\\data\\local\\someDir\\PP\\COLPEX\\COLPEX_16a_pj001.pp'
expected=('file', 'C:\\data\\local\\someDir\\PP\\COLPEX\\COLPEX_16a_pj001.pp')
actual=('file', 'C:\\data\\local\\someDir\\PP\\COLPEX\\COLPEX_16a_pj001.pp')

uri='file:///data/local/someDir/PP/COLPEX/COLPEX_16a_pj001.pp'
expected=('file', '/data/local/someDir/PP/COLPEX/COLPEX_16a_pj001.pp')
actual=('file', '/data/local/someDir/PP/COLPEX/COLPEX_16a_pj001.pp')

uri='http://www.somehost.com:8080/resource/thing.grib'
expected=('http', '/www.somehost.com:8080/resource/thing.grib')
actual=('http', '/www.somehost.com:8080/resource/thing.grib')

uri='/data/local/someDir/2013-11-25T13:49:17.632797'
expected=('file', '/data/local/someDir/2013-11-25T13:49:17.632797')
actual=('file', '/data/local/someDir/2013-11-25T13:49:17.632797')
```

Reply:
Are we surprised that the tests were ever passing then? I'm not sure I understand why they passed before.

Reply:
What actionable thing did you want me to do here?

Anyways, out of curiosity I checked out the `iris` v3.1.0 tag, rebuilt the associated conda environment from its lock file, and ran this test with the same diagnostics, and this is what I got:
```shell
uri='/data/local/someDir/PP/COLPEX/COLPEX_16a_pj001.pp'
pair=('file', '/data/local/someDir/PP/COLPEX/COLPEX_16a_pj001.pp')
actual=('file', '/data/local/someDir/PP/COLPEX/COLPEX_16a_pj001.pp')

uri='C:\\data\\local\\someDir\\PP\\COLPEX\\COLPEX_16a_pj001.pp'
pair=('file', 'C:\\data\\local\\someDir\\PP\\COLPEX\\COLPEX_16a_pj001.pp')
actual=('file', 'C:\\data\\local\\someDir\\PP\\COLPEX\\COLPEX_16a_pj001.pp')

uri='file:///data/local/someDir/PP/COLPEX/COLPEX_16a_pj001.pp'
pair=('file', '///data/local/someDir/PP/COLPEX/COLPEX_16a_pj001.pp')
actual=('file', '///data/local/someDir/PP/COLPEX/COLPEX_16a_pj001.pp')

uri='http://www.somehost.com:8080/resource/thing.grib'
pair=('http', '//www.somehost.com:8080/resource/thing.grib')
actual=('http', '//www.somehost.com:8080/resource/thing.grib')

uri='/data/local/someDir/2013-11-25T13:49:17.632797'
pair=('file', '/data/local/someDir/2013-11-25T13:49:17.632797')
actual=('file', '/data/local/someDir/2013-11-25T13:49:17.632797')
```
i.e., the behaviour of `iris.io.decode_uri` has changed, hence has the expected output.

Reply:
The behaviour that changed is exactly this...
```python
>>> from pathlib import Path
>>> p = Path("http://www.somehost.com:8080/resource/thing.grib")
>>> p
PosixPath('http:/www.somehost.com:8080/resource/thing.grib')
>>> str(p)
'http:/www.somehost.com:8080/resource/thing.grib'
>>> 
```
Note that the `//` has been collapsed down to a `/`, hence why the regex within `iris.io.decode_uri` (which hasn't changed) gives the result that it now gives.

Reply:
We should check the implications of this...